In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw8.ipynb")

# CPSC 330 - Applied Machine Learning 

## Homework 8: Time series
**Due date: See the [Calendar](https://calendar.google.com/calendar/u/0/embed?src=7a04205ae91b85e82ebc74daddbf2933c6b6723b81abb966f0e69c66a996c43b@group.calendar.google.com&ctz=America/Vancouver&pli=1).**

## Imports

In [2]:
from hashlib import sha1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, OneHotEncoder

from sklearn.linear_model import Ridge
from sklearn.ensemble import RandomForestRegressor

from sklearn.metrics import r2_score

## Submission instructions
<hr>
rubric={points:4}

You will receive marks for correctly submitting this assignment. 

Follow the homework instructions below and at the end of this file. 
There are also detailed [homework submission instructions](https://github.com/UBC-CS/cpsc330-2023s/blob/main/docs/homework_instructions.md) on github.

- **You may work on this assignment in a group (group size <= 4) and submit your assignment as a group.** 
- Below are some instructions on working as a group.  
    - The maximum group size is 4. 
    - You can choose your own group members. 
    - Use group work as an opportunity to collaborate and learn new things from each other. 
    - Be respectful to each other and make sure you understand all the concepts in the assignment well. 
    - It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. [Here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members) are some instructions on adding group members in Gradescope.  
- Upload the .ipynb file to Gradescope.
- **If the .ipynb file is too big or doesn't render on Gradescope for some reason, also upload a pdf or html in addition to the .ipynb.** 
- Make sure that your plots/output are rendered properly in Gradescope.

<br><br>

## Exercise 1: time series prediction

In this exercise we'll be looking at a [dataset of avocado prices](https://www.kaggle.com/neuromusic/avocado-prices). You should start by downloading the dataset. We will be forcasting average avocado price for the next week. 

In [3]:
df = pd.read_csv("data/avocado.csv", parse_dates=["Date"], index_col=0)
df.head()

Date  AveragePrice  Total Volume     4046       4225    4770  \
0 2015-12-27          1.33      64236.62  1036.74   54454.85   48.16   
1 2015-12-20          1.35      54876.98   674.28   44638.81   58.33   
2 2015-12-13          0.93     118220.22   794.70  109149.67  130.50   
3 2015-12-06          1.08      78992.15  1132.00   71976.41   72.58   
4 2015-11-29          1.28      51039.60   941.48   43838.39   75.78   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
1     9505.56     9408.07       97.49          0.0  conventional  2015  Albany  
2     8145.35     8042.21      103.14          0.0  conventional  2015  Albany  
3     5811.16     5677.40      133.76          0.0  conventional  2015  Albany  
4     6183.95     5986.26      197.69          0.0  conventional  2015  Albany

In [4]:
df.shape

(18249, 13)

In [5]:
df["Date"].min()

Timestamp('2015-01-04 00:00:00')

In [6]:
df["Date"].max()

Timestamp('2018-03-25 00:00:00')

It looks like the data ranges from the start of 2015 to March 2018 (~2 years ago), for a total of 3.25 years or so. Let's split the data so that we have a 6 months of test data.

In [7]:
split_date = '20170925'
df_train = df[df["Date"] <= split_date]
df_test  = df[df["Date"] >  split_date]

In [8]:
assert len(df_train) + len(df_test) == len(df)

<br><br>

<!-- BEGIN QUESTION -->

### 1.1 How many time series? 
rubric={points:4}

In the [Rain in Australia](https://www.kaggle.com/datasets/jsphyg/weather-dataset-rattle-package) dataset from lecture, we had different measurements for each Location. How about this avocado sales dataset? For which categorical feature(s), if any, do we have separate measurements? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.1
    
</div>

_Points:_ 4

In [9]:
print(df_train.columns)

Index(['Date', 'AveragePrice', 'Total Volume', '4046', '4225', '4770',
       'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'type', 'year',
       'region'],
      dtype='object')


Looking at the dataset features, there is one feature that sticks out as being useful to separate on, which would be "region". However, it might also be useful to look at "type". According to Kaggle, "type" is either conventional or organic, which tends to differ in prices.

For example, an avocado of type A in region 1 would be priced differently than an avocado of type B in region 1, or type A in region 2. 

In [10]:
print(df_train["region"].unique())
print(len(df_train["region"].unique()))

['Albany' 'Atlanta' 'BaltimoreWashington' 'Boise' 'Boston'
 'BuffaloRochester' 'California' 'Charlotte' 'Chicago' 'CincinnatiDayton'
 'Columbus' 'DallasFtWorth' 'Denver' 'Detroit' 'GrandRapids' 'GreatLakes'
 'HarrisburgScranton' 'HartfordSpringfield' 'Houston' 'Indianapolis'
 'Jacksonville' 'LasVegas' 'LosAngeles' 'Louisville' 'MiamiFtLauderdale'
 'Midsouth' 'Nashville' 'NewOrleansMobile' 'NewYork' 'Northeast'
 'NorthernNewEngland' 'Orlando' 'Philadelphia' 'PhoenixTucson'
 'Pittsburgh' 'Plains' 'Portland' 'RaleighGreensboro' 'RichmondNorfolk'
 'Roanoke' 'Sacramento' 'SanDiego' 'SanFrancisco' 'Seattle'
 'SouthCarolina' 'SouthCentral' 'Southeast' 'Spokane' 'StLouis' 'Syracuse'
 'Tampa' 'TotalUS' 'West' 'WestTexNewMexico']
54


In [11]:
print(df_train["type"].unique())

['conventional' 'organic']


In [12]:
# Some code inspired from Lecture 19 notes

region_df = df_train.sort_values(by=["region"])
region_df.head()

Date  AveragePrice  Total Volume     4046      4225   4770  \
0 2015-12-27          1.33      64236.62  1036.74  54454.85  48.16   
5 2016-11-20          1.93       1862.70   122.63    197.39   0.00   
6 2016-11-13          2.00       2084.37    75.80    191.07   0.00   
7 2016-11-06          1.93       2674.00    60.08    194.11   0.00   
8 2016-10-30          1.96       1326.80    57.66     93.41   0.00   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     8696.87     8603.62       93.25          0.0  conventional  2015  Albany  
5     1542.68     1542.68        0.00          0.0       organic  2016  Albany  
6     1817.50     1817.50        0.00          0.0       organic  2016  Albany  
7     2419.81     2419.81        0.00          0.0       organic  2016  Albany  
8     1175.73     1175.73        0.00          0.0       organic  2016  Albany

From just this short example above, we can see the AveragePrices of a few examples. Although they are all from the region Albany, the organic avocados are quite similar in price, while the conventional avocados are quite a bit cheaper. As a result, it might be useful to have a separate timeseries for all combinations region and type of avocados.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.2 Equally spaced measurements? 
rubric={points:4}

In the Rain in Australia dataset, the measurements were generally equally spaced but with some exceptions. How about with this dataset? Justify your answer by referencing the dataset.

<div class="alert alert-warning">

Solution_1.2
    
</div>

_Points:_ 4

_Type your answer here, replacing this text._

In [13]:
df_train.sort_values(['region', 'type', 'Date']).head(15)

Date  AveragePrice  Total Volume     4046      4225    4770  \
51 2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
50 2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
49 2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
48 2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
47 2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
46 2015-02-08          0.99      51253.97  1357.37  39111.81  163.25   
45 2015-02-15          1.06      41567.62   986.66  30045.51  222.42   
44 2015-02-22          1.07      45675.05  1088.38  35056.13  151.00   
43 2015-03-01          0.99      55595.74   629.46  45633.34  181.49   
42 2015-03-08          1.07      40507.36   795.68  30370.64  159.05   
41 2015-03-15          1.11      43045.79  2128.26  30447.17   99.67   
40 2015-03-22          1.12      46346.85  2141.83  34313.56  141.80   
39 2015-03-29          1.02      67799.08  1402.28  58623.22   89.50   
38 2015-04-05          1.16      47362.13   961.77  35577.66   93.76   
37 2015-04-12          1.13      48364.29   864.27  30374.15   21.50   

    Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
51     9716.46     9186.93      529.53         0.00  conventional  2015   
50     8424.77     8036.04      388.73         0.00  conventional  2015   
49    11921.05    11651.09      269.96         0.00  conventional  2015   
48    10845.82    10103.35      742.47         0.00  conventional  2015   
47     9323.18     9170.82      152.36         0.00  conventional  2015   
46    10621.54    10113.10      508.44         0.00  conventional  2015   
45    10313.03     9979.87      333.16         0.00  conventional  2015   
44     9379.54     9000.16      379.38         0.00  conventional  2015   
43     9151.45     8986.06      165.39         0.00  conventional  2015   
42     9181.99     8827.55      354.44         0.00  conventional  2015   
41    10370.69     9989.59      381.10         0.00  conventional  2015   
40     9749.66     9252.60      497.06         0.00  conventional  2015   
39     7684.08     7208.49      475.59         0.00  conventional  2015   
38    10728.94     9869.16      755.61       104.17  conventional  2015   
37    17104.37    16438.49      665.88         0.00  conventional  2015   

    region  
51  Albany  
50  Albany  
49  Albany  
48  Albany  
47  Albany  
46  Albany  
45  Albany  
44  Albany  
43  Albany  
42  Albany  
41  Albany  
40  Albany  
39  Albany  
38  Albany  
37  Albany

Seems like measurements are made every seven days. 
Let's sort descending to check if the pattern holds.

In [14]:
df_train.sort_values(by=['region', 'type', 'Date'], ascending=False).head(15)

Date  AveragePrice  Total Volume     4046     4225     4770  \
14 2017-09-24          2.26       9528.64  1545.34  2234.23     0.00   
15 2017-09-17          2.36      10464.29  1845.14  2819.17     0.00   
16 2017-09-10          2.38      11857.31  1562.10  4565.41     0.00   
17 2017-09-03          2.39       7657.47   927.27  4056.73     0.00   
18 2017-08-27          2.50      16137.93  2616.96  3672.96     0.00   
19 2017-08-20          2.43      10788.29  1665.31  3993.41     0.00   
20 2017-08-13          2.01       8435.52  1182.39  3688.93    11.24   
21 2017-08-06          1.78      15242.53  2134.24  4592.66   331.60   
22 2017-07-30          1.67      17633.69  2618.13  3169.61  1056.54   
23 2017-07-23          1.60      19144.32  1832.11  4752.03  1914.09   
24 2017-07-16          1.60      19126.81  1763.86  4382.01  2211.74   
25 2017-07-09          1.65      18244.09  2295.76  3701.18    38.67   
26 2017-07-02          1.59      20486.25  2454.99  3470.63     0.00   
27 2017-06-11          1.66      21170.11  3042.11  3588.46     0.00   
28 2017-06-04          1.59      25886.30  3305.88  3706.22     0.00   

    Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
14     5749.07     5722.40       26.67          0.0  organic  2017   
15     5799.98     5796.65        3.33          0.0  organic  2017   
16     5729.80     5719.96        9.84          0.0  organic  2017   
17     2673.47     2629.18       44.29          0.0  organic  2017   
18     9848.01     9816.58       31.43          0.0  organic  2017   
19     5129.57     5010.27      119.30          0.0  organic  2017   
20     3552.96     3403.12      149.84          0.0  organic  2017   
21     8184.03     8171.22       12.81          0.0  organic  2017   
22    10789.41    10733.97       55.44          0.0  organic  2017   
23    10646.09    10577.95       68.14          0.0  organic  2017   
24    10769.20    10652.21      116.99          0.0  organic  2017   
25    12208.48    11637.39      571.09          0.0  organic  2017   
26    14560.63    10390.44     4170.19          0.0  organic  2017   
27    14539.54     7757.19     6782.35          0.0  organic  2017   
28    18874.20     8654.49    10219.71          0.0  organic  2017   

              region  
14  WestTexNewMexico  
15  WestTexNewMexico  
16  WestTexNewMexico  
17  WestTexNewMexico  
18  WestTexNewMexico  
19  WestTexNewMexico  
20  WestTexNewMexico  
21  WestTexNewMexico  
22  WestTexNewMexico  
23  WestTexNewMexico  
24  WestTexNewMexico  
25  WestTexNewMexico  
26  WestTexNewMexico  
27  WestTexNewMexico  
28  WestTexNewMexico

To be absolutely sure, we can write some code to check for us:

In [15]:
types = ['conventional', 'organic']
for col in df_train["region"].unique():
    region = df_train[df_train['region'] == col]
    for t in types:
        type_df = region[region['type'] == t].sort_values("Date")
        time_diff = type_df['Date'].diff()
        if time_diff.dt.days.nunique() != 1:
            print(f"Rows in the date column of {col} for {t} avocados are not equally spaced.")

Rows in the date column of WestTexNewMexico for organic avocados are not equally spaced.


In [16]:
region = df_train[df_train['region'] == "WestTexNewMexico"]
type_df = region[region['type'] == "organic"].sort_values("Date")
time_diff = type_df['Date'].diff()
print(time_diff.value_counts())

7 days     137
14 days      1
21 days      1
Name: Date, dtype: int64


From the code, it looks like we do have equally spaced dates (7 days) with the exception of the region of "WestTextNewMexico", which has a couple of dates spaced at 14 and 21 days instead of the usual 7 days.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 1.3 Interpreting regions 
rubric={points:4}

In the Rain in Australia dataset, each location was a different place in Australia. For this dataset, look at the names of the regions. Do you think the regions are also all distinct, or are there overlapping regions? Justify your answer by referencing the data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

_Points:_ 4

Let's take a look at the regions first.

In [17]:
print(df_train["region"].unique())

['Albany' 'Atlanta' 'BaltimoreWashington' 'Boise' 'Boston'
 'BuffaloRochester' 'California' 'Charlotte' 'Chicago' 'CincinnatiDayton'
 'Columbus' 'DallasFtWorth' 'Denver' 'Detroit' 'GrandRapids' 'GreatLakes'
 'HarrisburgScranton' 'HartfordSpringfield' 'Houston' 'Indianapolis'
 'Jacksonville' 'LasVegas' 'LosAngeles' 'Louisville' 'MiamiFtLauderdale'
 'Midsouth' 'Nashville' 'NewOrleansMobile' 'NewYork' 'Northeast'
 'NorthernNewEngland' 'Orlando' 'Philadelphia' 'PhoenixTucson'
 'Pittsburgh' 'Plains' 'Portland' 'RaleighGreensboro' 'RichmondNorfolk'
 'Roanoke' 'Sacramento' 'SanDiego' 'SanFrancisco' 'Seattle'
 'SouthCarolina' 'SouthCentral' 'Southeast' 'Spokane' 'StLouis' 'Syracuse'
 'Tampa' 'TotalUS' 'West' 'WestTexNewMexico']


Just from a glance, you can see there's some overlap. 

For example, there are both states and cities within it.
- California and Sacramento, SanFrancisco, San Diego, LosAngeles, etc

There are regions that also overlap with cities.
- Boston and Northeast and NorthernNewEngland, etc

And of course, there's the region "TotalUS" that overlaps with everything.

<!-- END QUESTION -->

<br><br>

We will use the entire dataset despite any location-based weirdness uncovered in the previous part.

We will be trying to forecast the avocado price. The function below is adapted from [Lecture 19](https://github.com/UBC-CS/cpsc330-2023s/blob/main/lectures/19_time-series.ipynb), with some improvements.

In [18]:
def create_lag_feature(df, orig_feature, lag, groupby, new_feature_name=None, clip=False):
    """
    Creates a new feature that's a lagged version of an existing one.
    
    NOTE: assumes df is already sorted by the time columns and has unique indices.
    
    Parameters
    ----------
    df : pandas.core.frame.DataFrame
        The dataset.
    orig_feature : str
        The column name of the feature we're copying
    lag : int
        The lag; negative lag means values from the past, positive lag means values from the future
    groupby : list
        Column(s) to group by in case df contains multiple time series
    new_feature_name : str
        Override the default name of the newly created column
    clip : bool
        If True, remove rows with a NaN values for the new feature
    
    Returns
    -------
    pandas.core.frame.DataFrame
        A new dataframe with the additional column added.
        
    """
        
    if new_feature_name is None:
        if lag < 0:
            new_feature_name = "%s_lag%d" % (orig_feature, -lag)
        else:
            new_feature_name = "%s_ahead%d" % (orig_feature, lag)
    
    new_df = df.assign(**{new_feature_name : np.nan})
    for name, group in new_df.groupby(groupby):        
        if lag < 0: # take values from the past
            new_df.loc[group.index[-lag:],new_feature_name] = group.iloc[:lag][orig_feature].values
        else:       # take values from the future
            new_df.loc[group.index[:-lag], new_feature_name] = group.iloc[lag:][orig_feature].values
            
    if clip:
        new_df = new_df.dropna(subset=[new_feature_name])
        
    return new_df

We first sort our dataframe properly:

In [19]:
df_sort = df.sort_values(by=["region", "type", "Date"]).reset_index(drop=True)
df_sort

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   
18248 2018-03-25          1.62      15303.40  2325.30   2171.66    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   
18248    10806.44    10569.80      236.64          0.0       organic  2018   

                 region  
0                Albany  
1                Albany  
2                Albany  
3                Albany  
4                Albany  
...                 ...  
18244  WestTexNewMexico  
18245  WestTexNewMexico  
18246  WestTexNewMexico  
18247  WestTexNewMexico  
18248  WestTexNewMexico  

[18249 rows x 13 columns]

We then call `create_lag_feature`. This creates a new column in the dataset `AveragePriceNextWeek`, which is the following week's `AveragePrice`. We have set `clip=True` which means it will remove rows where the target would be missing.

In [20]:
df_hastarget = create_lag_feature(df_sort, "AveragePrice", +1, ["region", "type"], "AveragePriceNextWeek", clip=True)
df_hastarget

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18243 2018-02-18          1.56      17597.12  1892.05   1928.36    0.00   
18244 2018-02-25          1.57      18421.24  1974.26   2482.65    0.00   
18245 2018-03-04          1.54      17393.30  1832.24   1905.57    0.00   
18246 2018-03-11          1.56      22128.42  2162.67   3194.25    8.93   
18247 2018-03-18          1.56      15896.38  2055.35   1499.55    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18243    13776.71    13553.53      223.18          0.0       organic  2018   
18244    13964.33    13698.27      266.06          0.0       organic  2018   
18245    13655.49    13401.93      253.56          0.0       organic  2018   
18246    16762.57    16510.32      252.25          0.0       organic  2018   
18247    12341.48    12114.81      226.67          0.0       organic  2018   

                 region  AveragePriceNextWeek  
0                Albany                  1.24  
1                Albany                  1.17  
2                Albany                  1.06  
3                Albany                  0.99  
4                Albany                  0.99  
...                 ...                   ...  
18243  WestTexNewMexico                  1.57  
18244  WestTexNewMexico                  1.54  
18245  WestTexNewMexico                  1.56  
18246  WestTexNewMexico                  1.56  
18247  WestTexNewMexico                  1.62  

[18141 rows x 14 columns]

Our goal is to predict `AveragePriceNextWeek`. 

Let's split the data:

In [21]:
df_train = df_hastarget[df_hastarget["Date"] <= split_date]
df_test  = df_hastarget[df_hastarget["Date"] >  split_date]

<br><br>

### 1.4 `AveragePrice` baseline 
rubric={autograde:4}

Soon we will want to build some models to forecast the average avocado price a week in advance. Before we start with any ML though, let's try a baseline. Previously we used `DummyClassifier` or `DummyRegressor` as a baseline. This time, we'll do something else as a baseline: we'll assume the price stays the same from this week to next week. So, we'll set our prediction of "AveragePriceNextWeek" exactly equal to "AveragePrice", assuming no change. That is kind of like saying, "If it's raining today then I'm guessing it will be raining tomorrow". This simplistic approach will not get a great score but it's a good starting point for reference. If our model does worse that this, it must not be very good. 

Using this baseline approach, what $R^2$ do you get on the train and test data?

<div class="alert alert-warning">

Solution_1.4
    
</div>

_Points:_ 4

In [22]:
train_r2 = r2_score(df_train["AveragePriceNextWeek"], df_train["AveragePrice"])

In [23]:
test_r2 = r2_score(df_test["AveragePriceNextWeek"], df_test["AveragePrice"])

In [24]:
print(f"Train r2 score is {train_r2}")
print(f"Test r2 score is {test_r2}")

Train r2 score is 0.8285800937261841
Test r2 score is 0.7631780188583048


In [25]:
grader.check("q1.4")

q1.4 results: All test cases passed!

<br><br>

<!-- BEGIN QUESTION -->

### 1.5 Forecasting average avocado price
rubric={points:10}

Now that the baseline is done, let's build some models to forecast the average avocado price a week later. Experiment with a few approachs for encoding the date. Justify the decisions you make. Which approach worked best? Report your test score and briefly discuss your results.

Benchmark: you should be able to achieve $R^2$ of at least 0.79 on the test set. I got to 0.80, but not beyond that. Let me know if you do better!

Note: because we only have 2 splits here, we need to be a bit wary of overfitting on the test set. Try not to test on it a ridiculous number of times. If you are interested in some proper ways of dealing with this, see for example sklearn's [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html), which is like cross-validation for time series data.

<div class="alert alert-warning">

Solution_1.3
    
</div>

_Points:_ 10

In [26]:
# Some code inspired from Lecture 19 Notes

# Lets check for missing values
df_train.describe(include="all", datetime_is_numeric=True) #datetime_is_numeric=True to silence warning

Date  AveragePrice  Total Volume  \
count                           15441  15441.000000  1.544100e+04   
unique                            NaN           NaN           NaN   
top                               NaN           NaN           NaN   
freq                              NaN           NaN           NaN   
mean    2016-05-14 22:59:56.502817024      1.397126  8.415280e+05   
min               2015-01-04 00:00:00      0.440000  8.456000e+01   
25%               2015-09-06 00:00:00      1.080000  1.009291e+04   
50%               2016-05-15 00:00:00      1.360000  1.044206e+05   
75%               2017-01-22 00:00:00      1.660000  4.273913e+05   
max               2017-09-24 00:00:00      3.250000  6.103446e+07   
std                               NaN      0.409543  3.413913e+06   

                4046          4225          4770    Total Bags    Small Bags  \
count   1.544100e+04  1.544100e+04  1.544100e+04  1.544100e+04  1.544100e+04   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    2.915050e+05  2.987060e+05  2.438154e+04  2.269345e+05  1.739725e+05   
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     8.790700e+02  3.114340e+03  0.000000e+00  4.072320e+03  2.087380e+03   
50%     8.217170e+03  2.983921e+04  1.987000e+02  3.732293e+04  2.419380e+04   
75%     1.113626e+05  1.518530e+05  7.051080e+03  1.035380e+05  7.913568e+04   
max     2.274362e+07  2.047057e+07  2.546439e+06  1.629830e+07  1.256716e+07   
std     1.257911e+06  1.217028e+06  1.131981e+05  9.354825e+05  7.167709e+05   

          Large Bags    XLarge Bags          type          year  region  \
count   1.544100e+04   15441.000000         15441  15441.000000   15441   
unique           NaN            NaN             2           NaN      54   
top              NaN            NaN  conventional           NaN  Albany   
freq             NaN            NaN          7722           NaN     286   
mean    4.997246e+04    2989.486697           NaN   2015.909008     NaN   
min     0.000000e+00       0.000000           NaN   2015.000000     NaN   
25%     1.066700e+02       0.000000           NaN   2015.000000     NaN   
50%     2.348730e+03       0.000000           NaN   2016.000000     NaN   
75%     1.952399e+04     106.940000           NaN   2017.000000     NaN   
max     4.324231e+06  551693.650000           NaN   2017.000000     NaN   
std     2.215717e+05   17374.897311           NaN      0.792499     NaN   

        AveragePriceNextWeek  
count           15441.000000  
unique                   NaN  
top                      NaN  
freq                     NaN  
mean                1.401113  
min                 0.440000  
25%                 1.090000  
50%                 1.360000  
75%                 1.660000  
max                 3.250000  
std                 0.411514

Doesn't seem like any missing values.

In [27]:
# Dealing with dates
df_train["Date"].info

<bound method Series.info of 0       2015-01-04
1       2015-01-11
2       2015-01-18
3       2015-01-25
4       2015-02-01
           ...    
18218   2017-08-27
18219   2017-09-03
18220   2017-09-10
18221   2017-09-17
18222   2017-09-24
Name: Date, Length: 15441, dtype: datetime64[ns]>

Seems like the Date feature is already in Pandas Datetime. We can leave it as s then.

In [28]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15441 entries, 0 to 18222
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Date                  15441 non-null  datetime64[ns]
 1   AveragePrice          15441 non-null  float64       
 2   Total Volume          15441 non-null  float64       
 3   4046                  15441 non-null  float64       
 4   4225                  15441 non-null  float64       
 5   4770                  15441 non-null  float64       
 6   Total Bags            15441 non-null  float64       
 7   Small Bags            15441 non-null  float64       
 8   Large Bags            15441 non-null  float64       
 9   XLarge Bags           15441 non-null  float64       
 10  type                  15441 non-null  object        
 11  year                  15441 non-null  int64         
 12  region                15441 non-null  object        
 13  AveragePriceNext

Most of these are numbers. We'll need to be careful of "type" and "region". As we see here, we also have no missing data.

In [29]:
# We'll begin with separating features. Most features are numeric, except type and region which are categorical.  
# Our target is AveragePriceNextWeek, and we will also drop Date.

numeric_features = [
    "AveragePrice",
    "Total Volume",
    "4046",
    "4225",
    "4770",
    "Total Bags",
    "Small Bags",
    "Large Bags",
    "XLarge Bags",
    "year"
]
categorical_features = [
    "type",
    "region"
]
drop_features = ["Date"]
target = ["AveragePriceNextWeek"]

In [30]:
# Taken from Lecture 19 Notes

from sklearn.pipeline import make_pipeline
from sklearn.compose import ColumnTransformer, make_column_transformer

def preprocess_features(
    df_train,
    df_test,
    numeric_features,
    categorical_features,
    drop_features,
    target
):

    all_features = set(numeric_features + categorical_features + drop_features + target)
    if set(df_train.columns) != all_features:
        print("Missing columns", set(df_train.columns) - all_features)
        print("Extra columns", all_features - set(df_train.columns))
        raise Exception("Columns do not match")

    numeric_transformer = make_pipeline(
        SimpleImputer(strategy="median"), StandardScaler()
    )
    categorical_transformer = make_pipeline(
        SimpleImputer(strategy="constant", fill_value="missing"),
        OneHotEncoder(handle_unknown="ignore", sparse=False),
    )

    preprocessor = make_column_transformer(
        (numeric_transformer, numeric_features),
        (categorical_transformer, categorical_features),
        ("drop", drop_features),
    )
    preprocessor.fit(df_train)
    ohe_feature_names = (
        preprocessor.named_transformers_["pipeline-2"]
        .named_steps["onehotencoder"]
        .get_feature_names_out(categorical_features)
        .tolist()
    )
    new_columns = numeric_features + ohe_feature_names

    X_train_enc = pd.DataFrame(
        preprocessor.transform(df_train), index=df_train.index, columns=new_columns
    )
    X_test_enc = pd.DataFrame(
        preprocessor.transform(df_test), index=df_test.index, columns=new_columns
    )

    y_train = df_train["AveragePriceNextWeek"]
    y_test = df_test["AveragePriceNextWeek"]

    return X_train_enc, y_train, X_test_enc, y_test, preprocessor

In [31]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train,
    df_test,
    numeric_features,
    categorical_features,
    drop_features, 
    target
)

In [32]:
X_train_enc.head()

AveragePrice  Total Volume      4046      4225      4770  Total Bags  \
0     -0.432512     -0.234535 -0.229503 -0.222203 -0.214954   -0.232206   
1     -0.383676     -0.234440 -0.230948 -0.219448 -0.214272   -0.233587   
2     -0.554604     -0.233469 -0.231018 -0.219530 -0.214196   -0.229850   
3     -0.823205     -0.233283 -0.230996 -0.218170 -0.213945   -0.230999   
4     -0.994133     -0.225747 -0.230668 -0.196131 -0.213811   -0.232627   

   Small Bags  Large Bags  XLarge Bags      year  ...  region_SouthCarolina  \
0   -0.229907   -0.223154    -0.172063 -1.147053  ...                   0.0   
1   -0.231513   -0.223789    -0.172063 -1.147053  ...                   0.0   
2   -0.226469   -0.224325    -0.172063 -1.147053  ...                   0.0   
3   -0.228629   -0.222193    -0.172063 -1.147053  ...                   0.0   
4   -0.229930   -0.224856    -0.172063 -1.147053  ...                   0.0   

   region_SouthCentral  region_Southeast  region_Spokane  region_StLouis  \
0                  0.0               0.0             0.0             0.0   
1                  0.0               0.0             0.0             0.0   
2                  0.0               0.0             0.0             0.0   
3                  0.0               0.0             0.0             0.0   
4                  0.0               0.0             0.0             0.0   

   region_Syracuse  region_Tampa  region_TotalUS  region_West  \
0              0.0           0.0             0.0          0.0   
1              0.0           0.0             0.0          0.0   
2              0.0           0.0             0.0          0.0   
3              0.0           0.0             0.0          0.0   
4              0.0           0.0             0.0          0.0   

   region_WestTexNewMexico  
0                      0.0  
1                      0.0  
2                      0.0  
3                      0.0  
4                      0.0  

[5 rows x 66 columns]

### Linear Regression

In [33]:
def score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc):
    lr_pipe = make_pipeline(preprocessor, Ridge(max_iter=1000))
    lr_pipe.fit(df_train, y_train)
    print("Train score: {:.2f}".format(lr_pipe.score(df_train, y_train)))
    print("Test score: {:.2f}".format(lr_pipe.score(df_test, y_test)))
    lr_coef = pd.DataFrame(
        data=lr_pipe.named_steps["ridge"].coef_.flatten(),
        index=X_train_enc.columns,
        columns=["Coef"],
    )
    return lr_coef.sort_values(by="Coef", ascending=False)

In [34]:
score_lr_print_coeff(preprocessor, df_train, y_train, df_test, y_test, X_train_enc)

Train score: 0.85
Test score: 0.79


Coef
AveragePrice                0.328682
region_SanFrancisco         0.087368
region_HartfordSpringfield  0.084933
region_NewYork              0.067025
type_organic                0.049510
...                              ...
region_Denver              -0.045930
type_conventional          -0.049510
region_DallasFtWorth       -0.065811
region_SouthCentral        -0.065815
region_Houston             -0.075620

[66 rows x 1 columns]

We reach our benchmark of 0.79 with just Ridge. 

### POSIX Date

This time, instead of dropping the date, let's use the POSIX form of it.

In [35]:
date_train = df_train.Date.astype("int64").values.reshape(-1, 1) // 10 ** 9
date_test = df_test.Date.astype("int64").values.reshape(-1, 1) // 10 ** 9

In [36]:
# Let's remove the original date, and replace it with our new format.

df_train_posix = df_train.drop(columns=["Date"])
df_test_posix = df_test.drop(columns=["Date"])

df_train_posix.insert(0, "Date", date_train)
df_test_posix.insert(0, "Date", date_test)

In [37]:
df_test_posix.head()

Date  AveragePrice  Total Volume     4046      4225   4770  \
143  1506816000          1.69      71205.11  4411.02  57416.25  77.85   
144  1507420800          1.78      55368.61  3679.82  45843.75  42.63   
145  1508025600          1.65      73574.89  3383.35  63355.37  62.45   
146  1508630400          1.56      69704.09  3758.80  57340.30  35.48   
147  1509235200          1.67      69432.23  2959.76  57585.49  57.94   

     Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
143     9299.99     5069.66     4230.33          0.0  conventional  2017   
144     5802.41     2148.20     3654.21          0.0  conventional  2017   
145     6773.72     3882.02     2891.70          0.0  conventional  2017   
146     8569.51     5101.64     3467.87          0.0  conventional  2017   
147     8829.04     5050.91     3778.13          0.0  conventional  2017   

     region  AveragePriceNextWeek  
143  Albany                  1.78  
144  Albany                  1.65  
145  Albany                  1.56  
146  Albany                  1.67  
147  Albany                  1.62

Now that that's done, we'll go through the preprocessing again, except this time, we don't remove Date. It is instead a numeric feature.

In [38]:
numeric_features = [
    "Date",
    "AveragePrice",
    "Total Volume",
    "4046",
    "4225",
    "4770",
    "Total Bags",
    "Small Bags",
    "Large Bags",
    "XLarge Bags",
    "year"
]
categorical_features = [
    "type",
    "region"
]
drop_features = []
target = ["AveragePriceNextWeek"]

In [39]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train_posix,
    df_test_posix,
    numeric_features,
    categorical_features,
    drop_features, 
    target
)

In [40]:
X_train_enc.head()

Date  AveragePrice  Total Volume      4046      4225      4770  \
0 -1.719903     -0.432512     -0.234535 -0.229503 -0.222203 -0.214954   
1 -1.695677     -0.383676     -0.234440 -0.230948 -0.219448 -0.214272   
2 -1.671451     -0.554604     -0.233469 -0.231018 -0.219530 -0.214196   
3 -1.647225     -0.823205     -0.233283 -0.230996 -0.218170 -0.213945   
4 -1.622999     -0.994133     -0.225747 -0.230668 -0.196131 -0.213811   

   Total Bags  Small Bags  Large Bags  XLarge Bags  ...  region_SouthCarolina  \
0   -0.232206   -0.229907   -0.223154    -0.172063  ...                   0.0   
1   -0.233587   -0.231513   -0.223789    -0.172063  ...                   0.0   
2   -0.229850   -0.226469   -0.224325    -0.172063  ...                   0.0   
3   -0.230999   -0.228629   -0.222193    -0.172063  ...                   0.0   
4   -0.232627   -0.229930   -0.224856    -0.172063  ...                   0.0   

   region_SouthCentral  region_Southeast  region_Spokane  region_StLouis  \
0                  0.0               0.0             0.0             0.0   
1                  0.0               0.0             0.0             0.0   
2                  0.0               0.0             0.0             0.0   
3                  0.0               0.0             0.0             0.0   
4                  0.0               0.0             0.0             0.0   

   region_Syracuse  region_Tampa  region_TotalUS  region_West  \
0              0.0           0.0             0.0          0.0   
1              0.0           0.0             0.0          0.0   
2              0.0           0.0             0.0          0.0   
3              0.0           0.0             0.0          0.0   
4              0.0           0.0             0.0          0.0   

   region_WestTexNewMexico  
0                      0.0  
1                      0.0  
2                      0.0  
3                      0.0  
4                      0.0  

[5 rows x 67 columns]

In [41]:
score_lr_print_coeff(preprocessor, df_train_posix, y_train, df_test_posix, y_test, X_train_enc)

Train score: 0.85
Test score: 0.78


Coef
AveragePrice                0.324498
region_SanFrancisco         0.091838
region_HartfordSpringfield  0.089592
region_NewYork              0.070944
region_Philadelphia         0.052258
...                              ...
region_Denver              -0.047782
type_conventional          -0.052177
region_DallasFtWorth       -0.069429
region_SouthCentral        -0.070072
region_Houston             -0.079556

[67 rows x 1 columns]

### One-hot encoding the seasons

Fruits also grow in seasons, so it might make sense to encode date as seasons. 

In [42]:
df_train_month = df_train.assign(
    Month=df_train["Date"].apply(lambda x: x.month_name())
)  # x.month_name() to get the actual string
df_test_month = df_test.assign(Month=df_test["Date"].apply(lambda x: x.month_name()))

In [43]:
# From lecture 19
def get_season(month):
    # remember this is Australia
    WINTER_MONTHS = ["December", "January", "February"] 
    AUTUMN_MONTHS = ["September", "October", "November"]
    SUMMER_MONTHS = ["June", "July", "August"]
    SPRING_MONTHS = ["March", "April", "May"]
    if month in WINTER_MONTHS:
        return "Winter"
    elif month in AUTUMN_MONTHS:
        return "Autumn"
    elif month in SUMMER_MONTHS:
        return "Summer"
    else:
        return "Fall"

In [44]:
df_train_month

Date  AveragePrice  Total Volume     4046      4225    4770  \
0     2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1     2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2     2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3     2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4     2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   
...          ...           ...           ...      ...       ...     ...   
18218 2017-08-27          2.50      16137.93  2616.96   3672.96    0.00   
18219 2017-09-03          2.39       7657.47   927.27   4056.73    0.00   
18220 2017-09-10          2.38      11857.31  1562.10   4565.41    0.00   
18221 2017-09-17          2.36      10464.29  1845.14   2819.17    0.00   
18222 2017-09-24          2.26       9528.64  1545.34   2234.23    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  \
0         9716.46     9186.93      529.53          0.0  conventional  2015   
1         8424.77     8036.04      388.73          0.0  conventional  2015   
2        11921.05    11651.09      269.96          0.0  conventional  2015   
3        10845.82    10103.35      742.47          0.0  conventional  2015   
4         9323.18     9170.82      152.36          0.0  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
18218     9848.01     9816.58       31.43          0.0       organic  2017   
18219     2673.47     2629.18       44.29          0.0       organic  2017   
18220     5729.80     5719.96        9.84          0.0       organic  2017   
18221     5799.98     5796.65        3.33          0.0       organic  2017   
18222     5749.07     5722.40       26.67          0.0       organic  2017   

                 region  AveragePriceNextWeek      Month  
0                Albany                  1.24    January  
1                Albany                  1.17    January  
2                Albany                  1.06    January  
3                Albany                  0.99    January  
4                Albany                  0.99   February  
...                 ...                   ...        ...  
18218  WestTexNewMexico                  2.39     August  
18219  WestTexNewMexico                  2.38  September  
18220  WestTexNewMexico                  2.36  September  
18221  WestTexNewMexico                  2.26  September  
18222  WestTexNewMexico                  2.37  September  

[15441 rows x 15 columns]

In [45]:
df_train_month = df_train_month.assign(Season=df_train_month["Month"].apply(get_season)).drop("Month", axis=1).drop("Date", axis=1)
df_test_month = df_test_month.assign(Season=df_test_month["Month"].apply(get_season)).drop("Month", axis=1).drop("Date", axis=1)

In [46]:
df_train_month

AveragePrice  Total Volume     4046      4225    4770  Total Bags  \
0              1.22      40873.28  2819.50  28287.42   49.90     9716.46   
1              1.24      41195.08  1002.85  31640.34  127.12     8424.77   
2              1.17      44511.28   914.14  31540.32  135.77    11921.05   
3              1.06      45147.50   941.38  33196.16  164.14    10845.82   
4              0.99      70873.60  1353.90  60017.20  179.32     9323.18   
...             ...           ...      ...       ...     ...         ...   
18218          2.50      16137.93  2616.96   3672.96    0.00     9848.01   
18219          2.39       7657.47   927.27   4056.73    0.00     2673.47   
18220          2.38      11857.31  1562.10   4565.41    0.00     5729.80   
18221          2.36      10464.29  1845.14   2819.17    0.00     5799.98   
18222          2.26       9528.64  1545.34   2234.23    0.00     5749.07   

       Small Bags  Large Bags  XLarge Bags          type  year  \
0         9186.93      529.53          0.0  conventional  2015   
1         8036.04      388.73          0.0  conventional  2015   
2        11651.09      269.96          0.0  conventional  2015   
3        10103.35      742.47          0.0  conventional  2015   
4         9170.82      152.36          0.0  conventional  2015   
...           ...         ...          ...           ...   ...   
18218     9816.58       31.43          0.0       organic  2017   
18219     2629.18       44.29          0.0       organic  2017   
18220     5719.96        9.84          0.0       organic  2017   
18221     5796.65        3.33          0.0       organic  2017   
18222     5722.40       26.67          0.0       organic  2017   

                 region  AveragePriceNextWeek  Season  
0                Albany                  1.24  Winter  
1                Albany                  1.17  Winter  
2                Albany                  1.06  Winter  
3                Albany                  0.99  Winter  
4                Albany                  0.99  Winter  
...                 ...                   ...     ...  
18218  WestTexNewMexico                  2.39  Summer  
18219  WestTexNewMexico                  2.38  Autumn  
18220  WestTexNewMexico                  2.36  Autumn  
18221  WestTexNewMexico                  2.26  Autumn  
18222  WestTexNewMexico                  2.37  Autumn  

[15441 rows x 14 columns]

In [47]:
X_train_enc, y_train, X_test_enc, y_test, preprocessor = preprocess_features(
    df_train_month,
    df_test_month,
    [
    "AveragePrice",
    "Total Volume",
    "4046",
    "4225",
    "4770",
    "Total Bags",
    "Small Bags",
    "Large Bags",
    "XLarge Bags",
    "year"
],
    categorical_features + ["Season"],
    [], 
    target
)

In [48]:
X_train_enc.columns

Index(['AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags',
       'Small Bags', 'Large Bags', 'XLarge Bags', 'year', 'type_conventional',
       'type_organic', 'region_Albany', 'region_Atlanta',
       'region_BaltimoreWashington', 'region_Boise', 'region_Boston',
       'region_BuffaloRochester', 'region_California', 'region_Charlotte',
       'region_Chicago', 'region_CincinnatiDayton', 'region_Columbus',
       'region_DallasFtWorth', 'region_Denver', 'region_Detroit',
       'region_GrandRapids', 'region_GreatLakes', 'region_HarrisburgScranton',
       'region_HartfordSpringfield', 'region_Houston', 'region_Indianapolis',
       'region_Jacksonville', 'region_LasVegas', 'region_LosAngeles',
       'region_Louisville', 'region_MiamiFtLauderdale', 'region_Midsouth',
       'region_Nashville', 'region_NewOrleansMobile', 'region_NewYork',
       'region_Northeast', 'region_NorthernNewEngland', 'region_Orlando',
       'region_Philadelphia', 'region_PhoenixTucson', 'r

In [49]:
score_lr_print_coeff(preprocessor, df_train_month, y_train, df_test_month, y_test, X_train_enc)

Train score: 0.85
Test score: 0.79


Coef
AveragePrice                0.318218
region_SanFrancisco         0.098559
region_HartfordSpringfield  0.096385
region_NewYork              0.075819
region_Philadelphia         0.056151
...                              ...
region_Denver              -0.051734
type_conventional          -0.056059
region_SouthCentral        -0.073905
region_DallasFtWorth       -0.074257
region_Houston             -0.085293

[70 rows x 1 columns]

### Summary of Results

We tried 3 methods of encoding the date, and here are the results:

Original Pandas DateTime, dropped:
- 0.85 Train, 0.79 Test

POSIX
- 0.85 Train, 0.78 Test

Encoding as Season
- 0.85 Train, 0.79 Test.

The difference is very marginal, but the original dropped model and the model encoded with seasons performed better than the POSIX encoding method. The main difference between the dropped model and the seasons encoding was that the values for the coefficients were slightly different.

<!-- END QUESTION -->

<br><br><br><br>

## Exercise 2: very short answer questions

Each question is worth 2 points.

<!-- BEGIN QUESTION -->

### 2.1 Time series

rubric={points:4}

The following questions pertain to Lecture 19 on time series data:

1. Sometimes a time series has missing time points or, worse, time points that are unequally spaced in general. Give an example of a real world situation where the time series data would have unequally spaced time points.
2. In class we discussed two approaches to using temporal information: encoding the date as one or more features, and creating lagged versions of features. Which of these (one/other/both/neither) two approaches would struggle with unequally spaced time points? Briefly justify your answer.

<div class="alert alert-warning">

Solution_2.1
    
</div>

_Points:_ 4

1. Unequally spaced time points can happen quite often. For example, when data points happen without a set pattern or without warning.
- Purchases throughout the day/week/etc
- Bathroom visits throughout the day

2. 
- I believe lagged versions of features would struggle with unequally spaced time points. When feature lagging, we tend to use a set amount of time to lag, which wouldn't really work with unequal data points, since we would often not have points at the desired times.

- Encoding the date as a feature should not struggle as much with this, since features can be continuous without issues.

<!-- END QUESTION -->

<br><br>

<!-- BEGIN QUESTION -->

### 2.2 Survival analysis
rubric={points:6}

The following questions pertain to [Lecture 20](https://github.com/UBC-CS/cpsc330-2023s/blob/main/lectures/20_survival-analysis.ipynb) on survival analysis. We'll consider the use case of customer churn analysis.

1. What is the problem with simply labeling customers are "churned" or "not churned" and using standard supervised learning techniques?
2. Consider customer A who just joined last week vs. customer B who has been with the service for a year. Who do you expect will leave the service first: probably customer A, probably customer B, or we don't have enough information to answer?
3. If a customer's survival function is almost flat during a certain period, how do we interpret that?

<div class="alert alert-warning">

Solution_2.2
    
</div>

_Points:_ 6

1. The issue with simply labeling customers as "churned" or "not churned" is that it doesn't fully contextualize the information. For example, customers that will churn shortly after the data is recorded will be identified as "not churned", despite being quite different from customers that will not churn for a long time. We call the "not churned" data "censored" 

2. This could depend on the payment plan, so not enough information to answer. For example:
- Month-to-month subscription: There are plenty like customer A, who might join for just one month to try something out, then leave if it's not for them. Someone like customer B, who has stuck for the service for a long time, will probably stick around for longer.

- Year-long Contract: If this was the payment plan, there's a larger chance that customer B leaves, due to fulfilling their contract. Customer A would be less likely to leave, since they just started their contract and would require breaking it to leave.

3. A flat survival function means that, during that certain period, it is very likely for them to survive (not get churned).


<!-- END QUESTION -->

<br><br>

**PLEASE READ BEFORE YOU SUBMIT:** 

When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from "1" will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission. 
4. Make sure that the plots and output are rendered properly in your submitted file. If the .ipynb file is too big and doesn't render on Gradescope, also upload a pdf or html in addition to the .ipynb so that the TAs can view your submission on Gradescope. 

![](img/eva-well-done.png)